### ISSUE TITLE
constant folding: fold non-trivial constants

### ISSUE THREAD

**Issue by hugomg** [pallene-lang/pallene#531]
It would be nice if our constant folding pass could propagate things that aren't constant literals. For example:

``` lua
local PI = 3.14
local TAU = 2*PI

function m.f()
   -- something using TAU
end
```

**Comment by srijan-paul** [comment]
This kind of constant folding should come after constant propagation, no?

**Comment by hugomg** [comment]
I think it might want to interleave them. First we constant fold the outermost function. Then, we look at all the initUpvales inside it and see which ones are constant and propagate those constants into the inner functions. Then we repeat the process for the inner functions.

**Comment by srijan-paul** [comment]
Makes sense. Perhaps we do N passes until no more constants can be propagated / folded.
